### PR TITLE
bpo-28684: asyncio tests now handle PermissionError raised when bindi…

### PR DESCRIPTION
**Pull request by xdegaye**
…ng a unix socket

https://bugs.python.org/issue28684

### CODE REVIEW

**Comment by 1st1** [comment]
I would prefer to use `@unittest.skipIf(android)` approach instead.  `PermissionError` can occur on a normal Linux machine as a result of some regression in asyncio.  I don't want such cases to slip through our CI just because it sometimes happens on Android.

**Review by 1st1 — CHANGES_REQUESTED** [review]
I would prefer to use `@unittest.skipIf(android)` approach instead.  `PermissionError` can occur on a normal Linux machine as a result of some regression in asyncio.  I don't want such cases to slip through our CI just because it sometimes happens on Android.

**Comment by xdegaye** [comment]
@1st1 I understand your concern. Instead of ``@unittest.skipIf(android)`` I think it would be better to use a new decorator ``@support.skip_unless_bind_unix_socket`` to widen the scope of the changes:
* The root user on Android does not raise PermissionError, so it should be possible to run the asyncio test suite in that case without skipping those tests.
* Android runs now SElinux in enforcing mode and one of its policies restricts the access to unix sockets [1], so the problem is not Android specific and may impact other linux platforms depending on the way they are configured. As a confirmation of the fact that SElinux is involved, the remote command ``adb logcat`` prints this kind of AVC SElinux record whenever one of the asyncio tests fails with PermissionError:
```
11-23 15:36:11.859  2919  2919 W python  : type=1400 audit(0.0:20): avc: denied { create }
for name="tmpug8hwp0r" scontext=u:r:shell:s0 tcontext=u:object_r:shell_data_file:s0
tclass=sock_file permissive=0
```

[1] An interesting academic paper [The Misuse of Android Unix Domain Sockets and Security Implications](https://dl.acm.org/citation.cfm?id=2978297).

**Comment by 1st1** [comment]
SGTM, let's go for `@support.skip_unless_bind_unix_socket`

File: Lib/test/support/__init__.py [hunk 1]
```diff
@@ -2432,6 +2433,28 @@ def skip_unless_xattr(test):
...
+    """Decorator for tests requiring a functional bind() for unix sockets."""
+    if not hasattr(socket, 'AF_UNIX'):
+        return unittest.skip('No UNIX Sockets')(test)
+    global _bind_nix_socket_error
+    if _bind_nix_socket_error is None:
+        path = TESTFN + "can_bind_unix_socket"
+        with socket.socket(socket.AF_UNIX) as sock:
+            try:
+                sock.bind(path)
+                _bind_nix_socket_error = ''
```
Comment (1st1) [inline Lib/test/support/__init__.py]: `_bind_nix_socket_error = False`

Comment (xdegaye) [reply Lib/test/support/__init__.py]: Yes, much better.

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
LGTM

**Comment by xdegaye** [comment]
I am not sure that these changes need to be backported to 3.6.

**Pull request merged**